### DISCUSSION TITLE
Accessing Azure Key Vault

### DISCUSSION THREAD

**Discussion by mandalorianbob** [cake-build/cake#3397] [Q&A] [answered]
I need to download a pfx from Azure Key Vault. 

I would prefer to not do this in YAML, as I'm trying to move as much logic as possible into Cake.

Is there any kind of tool or extensions for accessing an Azure Key Vault and downloading a file from it?

--
Edit:
I should probably add that I'm trying to use Cake.Frosting for this, because kind of the whole point of moving from YAML to Cake is intellisense and being able to easily distribute the Tasks to other projects.

**Comment by nils-a** [comment, marked as answer]
Since downloading the certificate is possible using [Azure CLI](https://docs.microsoft.com/en-us/azure/key-vault/certificates/how-to-export-certificate?tabs=azure-cli#export-stored-certificates) it should be possible using [Cake.AzureCli](https://www.nuget.org/packages/Cake.AzureCli/).

I have not tested this, though.

In script
```cakescript
#addin nuget:?package=Cake.AzureCli&version=1.2.1

Task("Default")
.Does(() => {
   // probably need to authenticate, first...
   Az().Keyvault.Secret.Download(new AzKeyvaultSecretDownloadSettings 
   {
      File = "./my-cool-cert.pfx"
   });
});
```

In Frosting (add a reference to Cake.AzureCli NuGet):

```cs
[TaskName("Default")]
public class DefaultTask : FrostingTask<BuildContext>
{
    public override void Run(BuildContext context)
    {
        // probably need to authenticate, first...
        context.Az().Keyvault.Secret.Download(new AzKeyvaultSecretDownloadSettings
        {
            File = "./my-cool-cert.pfx"
        });
    }
}
```

As to your edit:

IntelliSense for Cake script is [available in VS Code](https://cakebuild.net/docs/integrations/editors/vscode/intellisense) and script files are as easily (maybe more so) distributable, as Frosting dlls: Both can be shared locally using a (shared) folder or published using NuGet packages. There was a nice article recently on that topic [from Lee Richardson](https://www.leerichardson.com/2021/02/he-one-thing-i-wish-id-known-before.html).

**Reply by mandalorianbob** [reply]
This worked! I misread "secret" as "service" as thought you were just misremembering the API, but it turns out that PFXs certs have to be downloaded through Secret. Yeesh, Azure.

But anyway, thanks! Helped a lot.
